[optim] Make casting to match params a hook

janeyx99 opened the pull request:
Moves the logic to casting state to match parameters into a hook so that users can choose to enable their hooks before or after the casting has happened.

With this, there is a little bit of redundancy of the id_map building and the check that the param groups are still aligned in length.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106725



cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

janeyx99 committed:
[optim] Make casting to match params a hook

[ghstack-poisoned]

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -720,44 +768,27 @@ def load_state_dict(self, state_dict: StateDict) -> None:
-        saved_lens = (len(g['params']) for g in saved_groups)
-        if any(p_len != s_len for p_len, s_len in zip(param_lens, saved_lens)):
-            raise ValueError("loaded state dict contains a parameter group "
-                             "that doesn't match the size of optimizer's group")
```
I removed the checks for this chunk above because we check it during the cast hook, but it may be safest to do the checks in both places.

Another option is to not check this in the hook.

albanD replied:
I guess there are two possible points here:
- Having it only here allows you to make sure this test will always run even if the user removes the cast hook and it will validate that what the user hooks generated is a valid state dict.
- Having it in the hook allows the user to "remove it" if they want this "broken" behavior for some reason.
- Have both so that we guarantee that both pre-hooks and post-hooks always generate valid state dict

Which one do you think is preferred here?

janeyx99 replied:
1. You're assuming we give users the handle to remove the preregistered hook 😛 which I don't yet. Thoughts on storing the returned handle as an attribute on the Optimizer class?
2. I am between either giving users freedom vs. having invariants on our state_dicts. Right now I am leaning towards having the checks in both to maintain the invariants. Do you have more thoughts?

albanD replied:
Given the current state of state dict, I would agree that for now, more checks are better than less. So I would vote on having it in both as well.

janeyx99 committed:
Update on "[optim] Make casting to match params a hook"

Moves the logic to casting state to match parameters into a hook so that users can choose to enable their hooks before or after the casting has happened.

With this, there is a little bit of redundancy of the id_map building and the check that the param groups are still aligned in length.




[ghstack-poisoned]

albanD commented on torch/optim/optimizer.py:
```diff
@@ -236,6 +236,52 @@ class Optimizer:
+                return value
+
+        # Cast state tensors to appropriate types in the state_dict
+        for k in state_dict['state'].keys():
```
```suggestion
        for k in state_dict['state']:
```

janeyx99 committed:
Update on "[optim] Make casting to match params a hook"

Moves the logic to casting state to match parameters into a hook so that users can choose to enable their hooks before or after the casting has happened.

With this, there is a little bit of redundancy of the id_map building and the check that the param groups are still aligned in length.




[ghstack-poisoned]

albanD commented on torch/optim/optimizer.py:
```diff
@@ -236,6 +236,51 @@ class Optimizer:
+                             "that doesn't match the size of optimizer's group")
+
+        # Update the state_dict with casted values
+        id_map = dict(zip(chain.from_iterable(g['params'] for g in saved_groups),
```
If you think this id_map and the associated testing is general, you could have a helper function to get it used both here and in load_state_dict.

janeyx99 replied:
Reasonable enough.

janeyx99 committed:
Update on "[optim] Make casting to match params a hook"

Moves the logic to casting state to match parameters into a hook so that users can choose to enable their hooks before or after the casting has happened.

With this, there is a little bit of redundancy of the id_map building and the check that the param groups are still aligned in length.




[ghstack-poisoned]

janeyx99 committed:
Update on "[optim] Make casting to match params a hook"

Moves the logic to casting state to match parameters into a hook so that users can choose to enable their hooks before or after the casting has happened.

With this, there is a little bit of redundancy of the id_map building and the check that the param groups are still aligned in length.




[ghstack-poisoned]

janeyx99 committed:
Update on "[optim] Make casting to match params a hook"

Moves the logic to casting state to match parameters into a hook so that users can choose to enable their hooks before or after the casting has happened.

With this, there is a little bit of redundancy of the id_map building and the check that the param groups are still aligned in length.




[ghstack-poisoned]

janeyx99 committed:
Update on "[optim] Make casting to match params a hook"

Moves the logic to casting state to match parameters into a hook so that users can choose to enable their hooks before or after the casting has happened.

With this, there is a little bit of redundancy of the id_map building and the check that the param groups are still aligned in length.




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

janeyx99 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1449,6 +1449,7 @@ def patch():
         # Note: we don't support sparsity, data-dependent control, or tracing through backwards
         excluded_opts = {torch.optim.SparseAdam, torch.optim.RAdam, torch.optim.LBFGS}
         for opt in optimizers:
+            opt.register_load_state_dict_pre_hook = disable(opt.register_load_state_dict_pre_hook)
```
With this change, the tests pass, but I'm not sure this is the best way to go about tracing optimizers @mlazos

janeyx99 committed:
Update on "[optim] Make casting to match params a hook"

Moves the logic to casting state to match parameters into a hook so that users can choose to enable their hooks before or after the casting has happened.

With this, there is a little bit of redundancy of the id_map building and the check that the param groups are still aligned in length.




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

janeyx99 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1449,6 +1449,13 @@ def patch():
+            # We disable `register_load_state_dict_pre_hook` to allow torch.compile to trace
+            # through the optimizer init without failing. Otherwise we get an irrecoverable graphbreak
+            # saying "torch.* op returned non-Tensor RemovableHandle call_function <class
+            # 'torch.utils.hooks.RemovableHandle'>"
```
See failures in https://hud.pytorch.org/pr/106987

albanD replied:
Sounds like a dynamo bug since graph breaking is fine here. We should open an issue with a small repro for the Dynamo team to look into it.

albanD reviewed (commented):
The test needs fixing. But SGTM otherwise

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -720,44 +792,21 @@ def load_state_dict(self, state_dict: StateDict) -> None:
-        groups = self.param_groups
-
         # Deepcopy as we write into saved_groups later to update state
         saved_groups = deepcopy(state_dict['param_groups'])
```
This is just the param_groups and not the whole state_dict (which includes state)

albanD replied:
Sounds ok then!

janeyx99 commented on test/optim/test_optim.py:
```diff
@@ -2064,6 +2064,27 @@ def test_load_state_dict_pre_post_hook(self):
+        cpu_state_dict = opt.state_dict()
+
+        param_cuda = torch.rand(2, 3, requires_grad=True, device="cuda")
+        param_cuda.grad = torch.rand_like(param_cuda)
```
this is valid 😅

janeyx99 committed:
Update on "[optim] Make casting to match params a hook"

Moves the logic to casting state to match parameters into a hook so that users can choose to enable their hooks before or after the casting has happened.

With this, there is a little bit of redundancy of the id_map building and the check that the param groups are still aligned in length.




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

janeyx99 commented:
@pytorchbot merge

janeyx99 committed:
Update on "[optim] Make casting to match params a hook"

Moves the logic to casting state to match parameters into a hook so that users can choose to enable their hooks before or after the casting has happened.

With this, there is a little bit of redundancy of the id_map building and the check that the param groups are still aligned in length.




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

janeyx99 committed:
Update on "[optim] Make casting to match params a hook"

Moves the logic to casting state to match parameters into a hook so that users can choose to enable their hooks before or after the casting has happened.

With this, there is a little bit of redundancy of the id_map building and the check that the param groups are still aligned in length.




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot revert -m "We acknowledge this is a huge risk because people do not remember to call super().__init__ from their Optimizer subclasses and so this will break lots of load_state_dict behavior" -c weird
